### DISCUSSION TITLE
I'm trying to create a very simple reverse-proxy but I'm unable to send body to the second server, why?

### DISCUSSION THREAD

**Discussion by frederikhors** [actix/actix-web#2847] [Q&A]
I'm using the below code (from [here](https://github.com/actix/examples/blob/master/http-proxy/src/main.rs)) to create a simple reverse-proxy for my application.

It works except when it must pass the body to the second server.

The second server gets no body (payload) from this code, why?

```rust
use actix_web::{
    error::{self},
    web, Error, HttpRequest, HttpResponse,
};

use crate::server::AppState;

pub async fn identity_server_proxy(
    req: HttpRequest,
    payload: web::Payload,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse, Error> {
    let forwarded_req = app_state
        .awc_client
        .request_from(
            format!(
                "{scheme}://{host}{path}{query}",
                scheme = "http",
                host = "localhost:3322",
                path = req.uri().path(),
                query = req.uri().query().unwrap_or(""),
            ),
            req.head(),
        )
        .no_decompress();

    let res = forwarded_req
        .send_stream(payload)
        .await
        .map_err(error::ErrorInternalServerError)?;

    let mut client_resp = HttpResponse::build(res.status());

    Ok(client_resp.streaming(res))
}
```

with `awc_client: awc::Client::default()` in `AppState`.

**Comment by Samielakkad** [comment]
One common trap in proxy code like this is that the forwarded request is built from the original request head, but the body stream is still a one-shot payload. If anything has already extracted/read the body before this handler, `send_stream(payload)` will forward an empty stream.

For a minimal proxy handler, keep the payload as `web::Payload` and pass it directly without touching it first:

```rust
pub async fn proxy(
    req: HttpRequest,
    payload: web::Payload,
    app_state: web::Data<AppState>,
) -> Result<HttpResponse, Error> {
    let target = format!(
        "http://localhost:3322{}{}",
        req.uri().path(),
        req.uri().query().map(|q| format!("?{q}")).unwrap_or_default(),
    );

    let mut forwarded = app_state.awc_client.request_from(target, req.head()).no_decompress();

    // Usually remove hop-by-hop headers when proxying.
    forwarded = forwarded
        .insert_header(("connection", "close"));

    let mut res = forwarded
        .send_stream(payload)
        .await
        .map_err(error::ErrorBadGateway)?;

    let mut client_resp = HttpResponse::build(res.status());
    for (name, value) in res.headers() {
        client_resp.insert_header((name.clone(), value.clone()));
    }

    Ok(client_resp.streaming(res))
}
```

Also check the upstream server method/content-type expectation. If the upstream route expects JSON/form data, make sure the incoming request has the same `Content-Type` and method. `request_from` copies most of that from the original head, but if a browser/client sent no body or the body was already consumed by an extractor/middleware, the upstream will see no payload.
